docs/source/coding_guidelines.md

```diff
@@ -0,0 +1,213 @@
+# Design and Coding Guidelines
+
+## Design Phase
+
+For non-trivial changes, please communicate your plans early on to reach maximum impact and to avoid duplicate work:
+
+* What are you planning to working on?
+* Why is it needed for the project?
+* Why is it needed in the repository?
+* Ask the rest of the team if they have already worked on it or thought about it. This helps to get to know about existing code fragments that can simplify your task, and integrate better with existing code.
+
+For non-trivial changes, we recommend to work with a team member to create a design for your work.
+Depending on the size of the task at hand, a design can be anywhere between a short descrption of
+your plan on Github, or a long document where you list your design options and invite feedback.
+
+Your design partner will later also be well positioned to review your PR because they are familiar with your plan.
+
+## Setting up your dev environment
+
+Please see the detailed instructions [here](developers.md).
+
+## Coding Guidelines
+
+### Naming
+
+Please follow the general Python rules for naming:
+
+* Variables, function and method names should use `snake_case` (lower case with under scores separating words)
+* Class names should follow `PascalCase` (first letter of each word capitalized).
+
+In addition:
+
+* To improve readability, functions or methods that return boolean values should follow a `is_...`, `has_...`, `use...` pattern, like `is_status_ok` instead of `ok`.
+
+### Static Analysis and Linting
+
+We use `flake8` as a linter, and `mypy` and pyright for static typechecking. Both tools run as part of the PR build, and must run without errors for a contribution to be accepted. `mypy` requires that all functions and methods carry type annotations, see [mypy documentation](https://mypy.readthedocs.io/en/latest/getting_started.html#function-signatures-and-dynamic-vs-static-typing)
+
+We highly recommend to run all those tools _before_ pushing the latest changes to a PR. If you have `make` installed, you can run both tools in one go via `make check` (from the repository root folder)
```

Comment:
Periods missing.

Instead of highly recommending running something manually, I think we should add the installation of `pre-commit` in the dev env doc, so these checks are run automatically. It's probably just replacing `conda env [...]` with `conda env [...] && pre-commit install`

Reply:
Tracking as https://github.com/microsoft/hi-ml/issues/342